src/openMVG/sfm/sfm_data_BA_ceres.cpp

```diff
@@ -180,27 +183,117 @@ bool Bundle_Adjustment_Ceres::Adjust(
     itIntrinsic != sfm_data.intrinsics.end(); ++itIntrinsic)
   {
-    const IndexT indexCam = itIntrinsic->first;
-
-    if (isValid(itIntrinsic->second->getType()))
+    const IndexT idIntrinsics = itIntrinsic->first;
+    if(intrinsicsUsage[idIntrinsics] == 0)
     {
-      map_intrinsics[indexCam] = itIntrinsic->second->getParams();
+      continue;
+    }
+    assert(isValid(itIntrinsic->second->getType()));
+    map_intrinsics[idIntrinsics] = itIntrinsic->second->getParams();
 
-      double * parameter_block = &map_intrinsics[indexCam][0];
-      problem.AddParameterBlock(parameter_block, map_intrinsics[indexCam].size());
-      if (!(refineOptions & BA_REFINE_INTRINSICS))
-      {
-        //set the whole parameter block as constant for best performance.
-        problem.SetParameterBlockConstant(parameter_block);
-      }
+    double * parameter_block = &map_intrinsics[idIntrinsics][0];
+    problem.AddParameterBlock(parameter_block, map_intrinsics[idIntrinsics].size());
+    if (!refineIntrinsics)
+    {
+      // Nothing to refine in the intrinsics,
+      // so set the whole parameter block as constant with better performances.
+      problem.SetParameterBlockConstant(parameter_block);
     }
     else
     {
-      std::cerr << "Unsupported camera type." << std::endl;
+      std::vector<int> vec_constant_params;
+      // Focal length
+      if(refineOptions & BA_REFINE_INTRINSICS_FOCAL)
+      {
+        // Refine the focal length
+        if(itIntrinsic->second->initialFocalLengthPix() > 0)
+        {
+          // If we have an initial guess, we only authorize a margin around this value.
+          assert(map_intrinsics[idIntrinsics].size() >= 1);
+          const unsigned int maxFocalErr = 0.2 * std::max(itIntrinsic->second->w(), itIntrinsic->second->h());
+          problem.SetParameterLowerBound(parameter_block, 0, (double)itIntrinsic->second->initialFocalLengthPix() - maxFocalErr);
+          problem.SetParameterUpperBound(parameter_block, 0, (double)itIntrinsic->second->initialFocalLengthPix() + maxFocalErr);
+        }
+        else // no initial guess
+        {
+          // We don't have an initial guess, but we assume that we use
+          // a converging lens, so the focal length should be positive.
+          problem.SetParameterLowerBound(parameter_block, 0, 0.0);
+        }
+      }
+      else
+      {
+        // Set focal length as constant
+        vec_constant_params.push_back(0);
+      }
+
+      // Optical center
+      if((refineOptions & BA_REFINE_INTRINSICS_OPTICALCENTER_ALWAYS) ||
+         ((refineOptions & BA_REFINE_INTRINSICS_OPTICALCENTER_IF_ENOUGH_DATA) && intrinsicsUsage[idIntrinsics] > 3)
+         )
+      {
+        // Refine optical center within 10% of the image size.
+        assert(map_intrinsics[idIntrinsics].size() >= 3);
+
+        // Add bounds to the principal point
+        problem.SetParameterLowerBound(parameter_block, 1, 0.45 * itIntrinsic->second->w());
```

Comment:
The idea is that it should be bounded around the center because we know that it's inside a window around the center. Many people are even keeping it at the center to avoid degenerated cases. So if we just limit the changes during each BA step, it can still create degenerated cases and we observed many degenerated cases due to the intrinsics!
So using a window of 10% of the image sounds reasonable and, at least, as proved to provide better results than a window of 20% on one problematic dataset.

I also observed a large improvement of the performances by adding these constraints on some datasets (up to 11x on one dataset).